processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -44,23 +54,62 @@
   private final IOPeon ioPeon;
   private final String filenameBase;
   private final ObjectStrategy<T> strategy;
-
+  private final int fileSizeLimit;
+  private final byte[] fileNameByteArray;
   private boolean objectsSorted = true;
   private T prevObject = null;
-
   private CountingOutputStream headerOut = null;
   private CountingOutputStream valuesOut = null;
-  int numWritten = 0;
+  private CountingOutputStream headerOutLong = null; // to hold header values as long and useful for version 2 GenericIndexed.
+  private long numWritten = 0;
+  private boolean requireMultipleFiles = false;
 
   public GenericIndexedWriter(
       IOPeon ioPeon,
       String filenameBase,
       ObjectStrategy<T> strategy
   )
   {
+    this(ioPeon, filenameBase, strategy, Integer.MAX_VALUE);
+  }
+
+  public GenericIndexedWriter(
+      IOPeon ioPeon,
+      String filenameBase,
+      ObjectStrategy<T> strategy,
+      int fileSizeLimit
+  )
+  {
     this.ioPeon = ioPeon;
     this.filenameBase = filenameBase;
     this.strategy = strategy;
+    this.fileSizeLimit = fileSizeLimit;
+    fileNameByteArray = filenameBase.getBytes();
+  }
+
+  public static String generateValueFileName(String fileNameBase, int fileNum)
+  {
+    return String.format("%s_value_%d", fileNameBase, fileNum);
+  }
+
+  public static String generateHeaderFileName(String fileNameBase)
+  {
+    return String.format("%s_header", fileNameBase);
+  }
+
+  private static void writeBytesIntoSmooshedChannel(
+      long numBytesToPutInFile,
+      final byte[] buffer,
+      final SmooshedWriter smooshChannel,
+      final InputStream is
+  )
+      throws IOException
+  {
+    while (numBytesToPutInFile > 0) {
+      int bytesRead = is.read(buffer, 0, Math.min(buffer.length, (int) numBytesToPutInFile));
```

Comment:
Unsafe cast to int, may lead to exception, suggested `Ints.saturatedCast(numBytesToPutInFile)`